### PR TITLE
Triton client

### PR DESCRIPTION
**Pull request by p-constant**
Add support for predict model served in Triton Inference Server:

````
from ultralytics import YOLO

model = YOLO("http://localhost:8000/yolov8n", task='detect') 

results = model("https://ultralytics.com/images/bus.jpg")[0]
````

While creating YOLO object user must specify "sheme+url+/model_name" as model and task.

### Summary
🌐🚀🔌

Added support for using remote models hosted on Triton Inference Server in the `ultralytics` package. The `Model` and `AutoBackend` classes can now handle Triton models as well as local models. A new module `ultralytics.utils.triton` was added with the `TritonRemoteModel` class that wraps the communication with the server.

> _Oh we are the `ultralytics` crew and we sail the AI seas_
> _We use the `Triton` server to run our models with ease_
> _We heave away and pull the rope, on the count of three_
> _We get our outputs from the `TritonRemoteModel` class, hooray!_

### Walkthrough
*  Add support for Triton Inference Server models as an alternative to local or hub models ([link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-cb649199881b55cff01fb5d1a216adf39d007be25e815089340a5b4f074cba00L12-R12), [link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-cb649199881b55cff01fb5d1a216adf39d007be25e815089340a5b4f074cba00L84-R93), [link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-cb649199881b55cff01fb5d1a216adf39d007be25e815089340a5b4f074cba00R108-R151), [link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-2e3ad5da4ba57eb68a9b5aaddf0c40c784d888b7619ed337ce31dfea4f6380f9L276-R278), [link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-2e3ad5da4ba57eb68a9b5aaddf0c40c784d888b7619ed337ce31dfea4f6380f9R388), [link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-2e3ad5da4ba57eb68a9b5aaddf0c40c784d888b7619ed337ce31dfea4f6380f9L492-R495), [link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-7f656783461e04167c1f5b4a6a05ad9b68af1023cea3924da3233ad6430f2ec4R1-R72))
  * Import `check_requirements` function from `ultralytics.utils.checks` to verify `tritonclient` dependency in `ultralytics.engine.model` ([link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-cb649199881b55cff01fb5d1a216adf39d007be25e815089340a5b4f074cba00L12-R12))
  * Add `else` clause to `Model.__init__` method to check if model argument is a Triton url and store model parameters and task in a dictionary ([link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-cb649199881b55cff01fb5d1a216adf39d007be25e815089340a5b4f074cba00L84-R93))
  * Add `Model.is_triton_model` static method to validate and parse Triton url, check model readiness and import `tritonclient` package ([link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-cb649199881b55cff01fb5d1a216adf39d007be25e815089340a5b4f074cba00R108-R151))
  * Replace `NotImplementedError` with code to create `TritonRemoteModel` object from weights dictionary in `AutoBackend.__init__` method ([link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-2e3ad5da4ba57eb68a9b5aaddf0c40c784d888b7619ed337ce31dfea4f6380f9L276-R278))
  * Convert `im` tensor to numpy array before passing it to model callable in `AutoBackend.predict` method ([link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-2e3ad5da4ba57eb68a9b5aaddf0c40c784d888b7619ed337ce31dfea4f6380f9R388))
  * Modify logic to handle dictionary as model argument and check for 'url' key in `AutoBackend._model_type` method ([link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-2e3ad5da4ba57eb68a9b5aaddf0c40c784d888b7619ed337ce31dfea4f6380f9L492-R495))
  * Add `ultralytics.utils.triton` module with `TritonRemoteModel` class definition that wraps `tritonclient` functionality and implements `__call__` method for inference ([link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-7f656783461e04167c1f5b4a6a05ad9b68af1023cea3924da3233ad6430f2ec4R1-R72))
* Add blank line for readability and consistency in `ultralytics.engine.model` ([link](https://github.com/ultralytics/ultralytics/pull/5102/files?diff=unified&w=0#diff-cb649199881b55cff01fb5d1a216adf39d007be25e815089340a5b4f074cba00L78-R78))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Added NVIDIA Triton Inference Server support and test, and improved utilities.

### 📊 Key Changes
- 🔨 Refactored utility imports to simplify and clean up code.
- ➕ Added comprehensive tests for the new Triton Inference Server functionality.
- 🆕 Implemented NVIDIA Triton Inference Server support in `ultralytics/nn/autobackend.py` and `ultralytics/engine/model.py`.
- 📝 Introduced new utility `is_valid_json` to check if a string is a valid JSON.
- 🛠 Fixed an issue in `ultralytics/models/fastsam/predict.py` by setting the number of classes to 1 for SAM models in non-max suppression.
- ✨ Created a new file `ultralytics/utils/triton.py` to handle communication and inference with Triton models.

### 🎯 Purpose & Impact
- 🚀 **Purpose**: 
  - To streamline code imports and usage.
  - To support NVIDIA's Triton Inference Server, which allows serving deep learning models over the network efficiently.
  - To make sure the SAM model inference handles class predictions correctly.
- 💥 **Impact**:
  - Developers can now integrate Triton server models with Ultralytics tools, potentially improving model deployment and scaling capabilities.
  - End-users gain access to high-performance inference capabilities provided by NVIDIA's infrastructure without detailed knowledge of its backend complexities.

### CODE REVIEW

**Comment by TheConstant3** [comment]
I have read the CLA Document and I sign the CLA

**Comment by glenn-jocher** [comment]
@TheConstant3 thanks for the PR! Is there a way to add triton deployment and inference to the tests? It looks like none of the new lines are covered, so there's no way to know if they are operating correctly, or for us to see if future PRs will break the added functionality.

**Comment by TheConstant3** [comment]
@glenn-jocher 
Can I create ```tests/test_triton.py``` file contains ```subprocess.run``` command for running tritonserver docker container? I have no idea how I can test triton deployment otherwise...

**Comment by glenn-jocher** [comment]
@TheConstant3 yes that's the idea! You can either create a tests/test_triton.py file if you need a lot of specialized code, or else if your tests can be scoped to a function you can simply create a new test function in tests/test_python.py

**Comment by glenn-jocher** [comment]
@TheConstant3 here's a test function for example that exports to ONNX and then runs inference with the ONNX model. If it's possible to do the same for Triton Server model that would be awesome!

https://github.com/ultralytics/ultralytics/blob/19c3314e68b47c8c5cb799f709868b083921067a/tests/test_python.py#L176-L179

**Review by p-constant — COMMENTED** [review]
refactor

**Comment by TheConstant3** [comment]
@TheConstant3 tests is written!

**Comment by glenn-jocher** [comment]
@TheConstant3 Great! Thank you for your contribution. Our team will review the tests and get back to you.

File: ultralytics/models/fastsam/predict.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ def postprocess(self, preds, img, orig_imgs):
                                     self.args.iou,
                                     agnostic=self.args.agnostic_nms,
                                     max_det=self.args.max_det,
-                                    nc=len(self.model.names),
```
[the remark this answers is not in the document]
Comment (glenn-jocher) [reply ultralytics/models/fastsam/predict.py:23]: @TheConstant3 or is this change just a simplification since SAM has no class predictions?

**Comment by glenn-jocher** [comment]
@TheConstant3 this line is showing a connection refused error in one test, though this seems to be only an occasional issue as other tests don't show this same issue. I'll re-run, but is there a way to make this test robust to this error?

Also do you have any updates on the `eval()` removal? We can't include the eval() line there due to downstream negative impacts to our security scores due to vulnerabilities introduced. See https://learn.snyk.io/lesson/malicious-code-injection/


Screenshot 2023-10-06 at 01 08 34

**Comment by glenn-jocher** [comment]
@TheConstant3 merging into dev branch for easier debugging. Please provide any assistance you can there on the above issues. Thank you!

**Pull request merged**